internals/overlord/logstate/manager_test.go

```diff
@@ -184,3 +201,199 @@ func (c *slowFlushingClient) Flush(ctx context.Context) error {
 		return nil
 	}
 }
+
+func (s *managerSuite) TestLabels(c *C) {
+	fakeClient := &labelStore{
+		labels:          map[string]map[string]string{},
+		notifySetLabels: make(chan struct{}, 2),
+	}
+
+	m := NewLogManager()
+	m.newGatherer = func(t *plan.LogTarget) (*logGatherer, error) {
+		return newLogGathererInternal(t, &logGathererOptions{
+			newClient: func(_ *plan.LogTarget) (logClient, error) { return fakeClient, nil },
+		})
+	}
+
+	svc1 := newTestService("svc1")
+	svc2 := newTestService("svc2")
+
+	pl := &plan.Plan{
+		Services: map[string]*plan.Service{
+			"svc1": svc1.config,
+			"svc2": svc2.config,
+		},
+		LogTargets: map[string]*plan.LogTarget{
+			"tgt1": {
+				Name:     "tgt1",
+				Type:     plan.LokiTarget,
+				Services: []string{"all"},
+				Labels: map[string]string{
+					"owner":   "user-$OWNER",
+					"address": "http://${IP}:${PORT}",
+				},
+			},
+		},
+	}
+
+	m.PlanChanged(pl)
+	checkGatherers(c, m.gatherers, map[string][]string{
+		"tgt1": nil,
+	})
+
+	m.ServiceStarted(svc1.config, &ServiceData{
+		Buffer: svc1.ringBuffer,
+		Env: map[string]string{
+			"OWNER": "alice",
+			"IP":    "103.2.51.6",
+			"PORT":  "3456",
+		},
+	})
+	m.ServiceStarted(svc2.config, &ServiceData{
+		Buffer: svc2.ringBuffer,
+		Env: map[string]string{
+			"IP":   "103.2.52.88",
+			"PORT": "9090",
+		},
+	})
+
+	fakeClient.wait(c, 1*time.Second)
+	fakeClient.wait(c, 1*time.Second)
+	c.Assert(fakeClient.labels, DeepEquals, map[string]map[string]string{
+		"svc1": {
+			"owner":   "user-alice",
+			"address": "http://103.2.51.6:3456",
+		},
+		"svc2": {
+			"owner":   "user-", // undefined env vars -> empty string
+			"address": "http://103.2.52.88:9090",
+		},
+	})
+
+	// If we change only the target's labels (with no change to the services),
+	// we still need to recalculate the client's labels.
+	pl.LogTargets["tgt1"].Labels["foo"] = "bar"
+	m.PlanChanged(pl)
+
+	// Wait for labels to be set
+	fakeClient.wait(c, 1*time.Second)
+	fakeClient.wait(c, 1*time.Second)
+	c.Assert(fakeClient.labels, DeepEquals, map[string]map[string]string{
+		"svc1": {
+			"owner":   "user-alice",
+			"address": "http://103.2.51.6:3456",
+			"foo":     "bar",
+		},
+		"svc2": {
+			"owner":   "user-",
+			"address": "http://103.2.52.88:9090",
+			"foo":     "bar",
+		},
+	})
+}
+
+// Fake logClient implementation which just stores the passed-in labels
+type labelStore struct {
+	labels map[string]map[string]string
+	// synchronise on this channel to avoid data races
+	notifySetLabels chan struct{}
+}
+
+func (c *labelStore) Add(_ servicelog.Entry) error {
+	return nil // no-op
+}
+
+func (c *labelStore) Flush(_ context.Context) error {
+	return nil // no-op
+}
+
+func (c *labelStore) SetLabels(serviceName string, labels map[string]string) {
+	c.labels[serviceName] = labels
+	c.notifySetLabels <- struct{}{}
+}
+
+// wait for labels to be set
+func (l *labelStore) wait(c *C, timeout time.Duration) {
```

Comment:
This name + args confused me at first, as I thought the caller was saying "wait 1s", which wouldn't be good in tests. What about we just hard-code the timeout in this method as it's always 1s, and rename the method to `waitLabels` or `waitLabelsSet`?